hudi-common/src/main/java/org/apache/parquet/schema/AvroSchemaRepair.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.parquet.schema;
+
+import org.apache.hudi.avro.AvroSchemaCache;
+import org.apache.hudi.avro.AvroSchemaUtils;
+
+import org.apache.avro.LogicalType;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class AvroSchemaRepair {
+  public static Schema repairLogicalTypes(Schema fileSchema, Schema tableSchema) {
+    Schema repairedSchema = repairAvroSchema(fileSchema, tableSchema);
+    if (repairedSchema != fileSchema) {
+      return AvroSchemaCache.intern(repairedSchema);
+    }
+    return fileSchema;
+  }
+
+  /**
+   * Performs schema repair on a schema, handling nullable unions.
+   */
+  private static Schema repairAvroSchema(Schema fileSchema, Schema tableSchema) {
+    // Always resolve nullable schemas first (returns unchanged if not a union)
+    Schema nonNullFileSchema = AvroSchemaUtils.resolveNullableSchema(fileSchema);
+    Schema nonNullTableSchema = AvroSchemaUtils.resolveNullableSchema(tableSchema);
+
+    // Perform repair on the non-null types
+    Schema nonNullRepairedSchema = repairAvroSchemaNonNull(nonNullFileSchema, nonNullTableSchema);
+
+    // If nothing changed, return the original schema
+    if (nonNullRepairedSchema == nonNullFileSchema) {
+      return fileSchema;
+    }
+
+    // If the original was a union, wrap the repaired schema back in a nullable union
+    if (fileSchema.getType() == Schema.Type.UNION) {
+      return AvroSchemaUtils.createNullableSchema(nonNullRepairedSchema);
```

Comment:
does this mean, every column is to be nullable after this point? 
will it not have any side effects? i.e. when repairing is needed, we end up converting all columns as nullable. where as when repair is not required, we do not convert and likely use whats incoming or table schema as is.